### DISCUSSION TITLE
Process for submitting a bug PR with test

### DISCUSSION THREAD

**Discussion by tonsoflaz2** [livewire/livewire#3494] [1. Help] [answered]
Hi,
I am looking for some best practices or guidelines or even specific steps to be able to contribute to this repo.
I know that Caleb has said the *ideal* bug submission or PR would start with adding a failing test (one that should work) so that everyone can see the problem without recreating their own environment. 

BUT, where are the current tests run? I presume the repo must have an extensive test suite that runs to make sure contributors didn't break anything, but I don't see it. Is that maybe an external repo that loads this and runs the tests?
Thank you for any advice!

**Comment by joshhanley** [comment, marked as answer]
@tonsoflaz2 have a look in the `tests` directory https://github.com/livewire/livewire/tree/master/tests

The tests are run using phpunit, and the suite consists of unit tests and Laravel dusk tests.

Hope this helps!

**Reply by tonsoflaz2** [reply]
Thank you, I'm not sure where I was looking that I didn't see the "tests" directory!

Addendum, part 2, continued:
1. I cloned the repo
2. I ran composer install
3. I ran vendor/bin/phpunit

It errors out with "Failed to connect to localhost port 9515: Connection refused"

I recognize this as a Dusk error, but shouldn't dusk be included in the livewire install?
Thank you for any guidance in advance, I am really just trying to remove all pain points in quickly fixing bugs or making PRs to the repo, for myself and maybe for future contributors!

**Reply by joshhanley** [reply]
@tonsoflaz2 no worries! 🙂

Yeah that is a dusk error, so the fact that you are getting it suggests dusk did actually install.

Dusk can be a bit fickle. I find opening a normal chrome browser window, then trying to run the tests sometimes works. Other times its the opposite, close all chrome instances and try to run tests.

But you may also have a driver mismatch, you can run `./vendor/bin/dusk-updater detect --auto-update` to update the dusk chromedriver it uses.

Livewire uses https://github.com/orchestral/testbench-dusk if you need further troubleshooting, it's worth checking that out. That is where the above command comes from.

Hope this helps!

**Reply by tonsoflaz2** [reply]
That was VERY helpful! I ran:
`./vendor/bin/dusk-updater detect --auto-update `
And it automatically detected the chromedriver was missing and installed the correct one.

After that, I ran the tests again:
`./vendor/bin/phpunit`

This time it went through many of the tests, opened Chrome, ran dusk, but then failed on this test:

> **From tests/Browser/Alpine/Transition/Test.php:75:**
>     73:             ->click('@'.$secondHook)
>     74:             ->pause(100)
>   75:             ->assertScript('document.querySelector(\'[dusk="outer"]\').style.disp                 
>  76:assertScript('document.querySelector(\'[dusk="inner"]\').style.disp    77:             ->. assertScript('document.querySelector(\'[dusk="inner"]\').style.opac
> JavaScript expression [return document.querySelector('[dusk="outer"]').style.display] misFailed asserting that two strings are equal.

Because it is one of the Alpine tests, I have to guess it has to do with an alpine version or missing alpine library?

Josh -- please know that I really appreciate your guidance and am NOT expecting a response here, but I want to document the process fully. Ideally, anyone should be able to clone the repo, composer install, then run the tests successfully as a first step to contributing. so if anyone has an idea why this test is failing, please let me know, and if I figure it out I will post it here!

**Reply by joshhanley** [reply]
@tonsoflaz2 awesome! Glad you got it working 🙂

So as I said before, dusk can be fickle. Sometimes you just need to re-run the test suite a couple of times and then it should pass.

But if you want confirmation, if you check the home page of the repo, and the last commit was green and all tests passed, then you can know it's just something funny happening locally with dusk.

![image](https://user-images.githubusercontent.com/882837/128931997-231bce73-e20c-4030-a035-c02a847b1a84.png)

Sometimes there might actually be a failing test on master, so worth checking just to make sure. No point trying to waste time investigating a failing test if you don't need to.

Hope this helps!
